llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8854,11 +8854,16 @@ ConstantRange llvm::computeConstantRange(const Value *V, bool ForSigned,
   } else if (auto *II = dyn_cast<IntrinsicInst>(V))
     CR = getRangeForIntrinsic(*II);
   else if (auto *SI = dyn_cast<SelectInst>(V)) {
+    ConstantRange CRTrue = computeConstantRange(
+        SI->getTrueValue(), ForSigned, UseInstrInfo, AC, CtxI, DT, Depth + 1);
+    ConstantRange CRFalse = computeConstantRange(
+        SI->getFalseValue(), ForSigned, UseInstrInfo, AC, CtxI, DT, Depth + 1);
+    CR = CRTrue.unionWith(CRFalse);
```

Comment:
code wise looks fine.
Maybe add a todo for (or just implement) trying to use select condition. I.e we have something like:
`select (icmp ule x, 10), x, 10`
we can bound to [0, 10].
Something like:
```
const APInt * C;
if(match(SelectCond, m_ICmp(Pred, m_Specific(SelectTrue), m_APInt(C))) {
    CRTrue = CRTrue.intersectWith(ConstantRange::makeExactICmpRegion(Pred), *C);
}
if(match(SelectCond, m_ICmp(Pred, m_Specific(SelectFalse), m_APInt(C))) {
   CRFalse = CRFalse.intersectWith(ConstantRange::makeExactICmpRegion(getInverse(Pred), *C));
}
```
You could get a bit more sophisticated and do non-constants of just `m_APInt(C)` and build a CR for the other operand.

Reply:
@goldsteinn I think `setLimitsForSelectPattern` does what you describe.

Reply:
not quite, but the logic i'm proposing looks like it belongs better there. I'll make a patch for that.

Reply:
Still think the logic belongs here actually, but feel free to commit w.o adding, I'll likely add a follow up shortly after.

Reply:
@goldsteinn Maybe I didn't understand what you mean. But I tried the following test without my patch:

```
define i1 @foo(i32 %x)  {
entry:
  %cmp0_ = icmp ule i32 %x, 10
  %select_ = select i1 %cmp0_, i32 %x, i32 10
  %cmp1_ = icmp ule i32 %select_, 10
  ret i1 %cmp1_
}
```

This simplifies to:

```
define i1 @foo(i32 %x) {
entry:
  ret i1 true
}
```

I put a print statement right after a call to `setLimitsForSelectPattern(*SI, Lower, Upper, IIQ)`, the return range is `[0, 11)`.
Can you please demonstrate what's missing with another test case?